torch/csrc/jit/tensorexpr/codegen.cpp

```diff
@@ -185,34 +186,45 @@ std::vector<std::pair<BufPtr, BufPtr>> AllocBufsWithMemReuse(
   return buf_allocs;
 }
 
-StmtPtr insertAllocFree(
-    std::vector<std::pair<BufPtr, BufPtr>>& buf_allocs,
-    StmtPtr stmt) {
-  BlockPtr b = to<Block>(stmt);
+void CodeGen::insertAllocFree(
+    std::vector<std::pair<BufPtr, BufPtr>>& buf_to_alloc_maps,
+    const bool pre_alloc) {
+  BlockPtr b = to<Block>(stmt_);
   if (!b) {
-    b = alloc<Block>(std::vector<StmtPtr>({stmt}));
+    b = alloc<Block>(std::vector<StmtPtr>({stmt_}));
   }
 
-  // Insert allocations and frees for temporary buffers at global scope.
-  for (auto rit = buf_allocs.rbegin(); rit != buf_allocs.rend(); ++rit) {
+  // Go through intermediate buffer list with reuse info to allocate these
+  // buffers.
+  for (auto rit = buf_to_alloc_maps.rbegin(); rit != buf_to_alloc_maps.rend();
+       ++rit) {
+    // Only allocate buffers that are mapped to selves.
     if (rit->first == rit->second) {
       BufPtr buf = rit->first;
-      b->prepend_stmt(alloc<Allocate>(buf));
-      b->append_stmt(alloc<Free>(buf));
+      if (pre_alloc) {
+        // Save buffers for buffer preallocation in TEK.
+        buffers_to_pre_alloc_.emplace_back(buf);
+        // Add preallocated buffers to args.
+        buffer_args_.emplace_back(BufHandle(buf));
```

Comment:
It seems like if `pre_alloc` is true, but the buffer has dynamic shapes, it won't be allocated at all. Here we would expect TEK to pre-allocate it, but TEK would reject to do that since shapes are not constants. Am I missing something? Do you mind adding a test to exercise this?

Reply:
That's a good point! Have fixed it and added a test.